timemachine/cpp/src/context.cu

```diff
@@ -272,77 +203,35 @@ std::array<std::vector<double>, 2> Context::multiple_steps_local(
         // Setup the flat bottom restraints
         bound_shell_restraint->set_params_device(std::vector<int>({num_row_indices, 3}), bond_params.data, stream);
         restraint_ptr->set_bonds_device(num_row_indices, restraints.data, stream);
-        // Invert to get the column indices
-        k_invert_indices<<<ceil_divide(N_, tpb), tpb, 0, stream>>>(N_, d_shell_idxs_inner.data);
-        gpuErrchk(cudaPeekAtLastError());
 
-        // Partition the col idxs to the front
-        cub::DevicePartition::If(
-            d_temp_storage_buffer.data,
-            temp_storage_bytes,
-            d_shell_idxs_inner.data,
-            d_col_idxs.data,
-            num_selected_buffer.data,
-            N_,
-            select_op,
-            stream);
-        gpuErrchk(cudaPeekAtLastError());
+        // Set the nonbonded potential to compute forces of free particles
+        set_nonbonded_potential_idxs(nonbonded_bp->potential, num_row_indices, d_row_idxs.data, stream);
 
-        // Invert to get back to the inner shell idxs
-        k_invert_indices<<<ceil_divide(N_, tpb), tpb, 0, stream>>>(N_, d_shell_idxs_inner.data);
+        // Invert to get column indices
+        k_invert_indices<<<ceil_divide(N_, tpb), tpb, 0, stream>>>(N_, d_free_indices.data);
         gpuErrchk(cudaPeekAtLastError());
 
-        nblist.set_idxs_device(num_col_indices, num_row_indices, d_col_idxs.data, d_row_idxs.data, stream);
-        int max_interactions = nblist.max_ixn_count();
-        // Build the neighborlist around the inner idxs to get the outer shell. Use the nonbonded potential's cutoff
-        // to ensure correctness and to avoid wasted computation (ie radius >> outer_cutoff).
-        nblist.build_nblist_device(N_, d_x_t_, d_box_t_, outer_cutoff, stream);
-
-        // Now reuse the d_col_idxs for the outer idxs, to reduce memory consumption
-        // Set the array to all N, which means it will be ignored as an idx
-        k_initialize_array<unsigned int><<<ceil_divide(N_, tpb), tpb, 0, stream>>>(N_, d_col_idxs.data, N_);
-        gpuErrchk(cudaPeekAtLastError());
-
-        k_unique_indices<<<ceil_divide(max_interactions, tpb), tpb, 0, stream>>>(
-            max_interactions, N_, nblist.get_ixn_atoms(), d_col_idxs.data);
-        gpuErrchk(cudaPeekAtLastError());
-
-        // Add the inner indices to the outer indices
-        k_unique_indices<<<ceil_divide(num_row_indices, tpb), tpb, 0, stream>>>(
-            num_row_indices, N_, d_row_idxs.data, d_col_idxs.data);
-        gpuErrchk(cudaPeekAtLastError());
```

Comment:
This looks like an improvement in safety -- I did not have a good grasp of when the "inner and outer shell" optimization was correct.

Reply:
Definitely much cleaner with the ixn group code